File: src/class036/C2_deque.py

from collections import deque
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def zigzagLevelOrder(self, root: TreeNode):
        if not root:
            return []

        results = []
        queue = deque([root])
        reverse = False

        while queue:
            size = len(queue)
            level = []
            # 计算起始位置和步长
            start = len(queue) - 1 if reverse else 0
            step = -1 if reverse else 1

            # 从 start 开始，根据 step 的正负值递增或递减 i
            for i, k in zip(range(start, start + size * step, step), range(size)):
                node = queue[i]
                level.append(node.val)

            # 进行下一层节点的追加
            new_queue = deque()
            while queue:
                node = queue.popleft()
                if node.left:
                    new_queue.append(node.left)
                if node.right:
                    new_queue.append(node.right)
            
            queue = new_queue
            results.append(level)
            reverse = not reverse

        return results

File: src/class036/C2_deque.cpp
   vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if (!root) {
            return {};
        }

        vector<vector<int>> results;
        deque<TreeNode*> queue;
        queue.push_back(root);
        bool reverse = false;

        while (!queue.empty()) {
            int size = queue.size();
            vector<int> level(size);

            // Calculate start position and step based on whether we are reversing
            int start = reverse ? size - 1 : 0;
            int step = reverse ? -1 : 1;

            for (int i = 0; i < size; ++i) {
                TreeNode* node = reverse ? queue.back() : queue.front();
                if (reverse) queue.pop_back(); else queue.pop_front();
                level[start + i * step] = node->val;
                
                // Enqueue child nodes in the order required for next level's processing order
                if (!reverse) {
                    if (node->left) queue.push_back(node->left);
                    if (node->right) queue.push_back(node->right);
                } else {
                    if (node->right) queue.push_front(node->right);
                    if (node->left) queue.push_front(node->left);
                }
            }
            
            results.push_back(move(level));
            reverse = !reverse;
        }
        
        return results;
    }